### PR TITLE
Interrupt search tree deapth when stop criterion is reached

### PR DESCRIPTION
**Pull request by phiedw**
**Please check if the PR fulfills these requirements** *(please use `'[x]'` to check the checkboxes, or submit the PR and then click the checkboxes)*
- [x] The commit message follows our guidelines
- [x] Tests for the changes have been added (for bug fixes / features)

**Does this PR already have an issue describing the problem ?** *If so, link to this issue using `'#XXX'` and skip the rest*
No

**What kind of change does this PR introduce?** *(Bug fix, feature, docs update, ...)*
Feature

**What is the current behavior?** *(You can also link to an open issue here)*
The search tree will finish optimizing all the leaves of the current depth before checking if the stop criterion is reached.

**What is the new behavior (if this is a feature change)?**
The search tree now stops as soon as a leaf reaches the stop criterion (the leaves "before" it are still allowed to finish and we select the first leaf that reached the stop criterion, to avoid reproductibility issues)

**Does this PR introduce a breaking change?** *(What changes might users need to make in their application due to this PR?)*
No

**Other information**:

(if any of the questions/checkboxes don't apply, please delete them entirely)

### CODE REVIEW

File: ra-optimisation/search-tree-rao/src/main/java/com/farao_community/farao/search_tree_rao/SearchTree.java [hunk 1]
```diff
@@ -373,9 +390,23 @@ private SensitivityComputer getSensitivityComputerForOptimizationBasedOn(FlowRes
...
+                    combinationFulfillingStopCriterion = Optional.of(networkActionCombination);
+                }
+            }
+            // special case: stop criterion has been reached
+            if (combinationFulfillingStopCriterion.isPresent()
+                && stopCriterionReached(leaf)
+                && orderNetworkActionCombinationsRandomly(networkActionCombination, combinationFulfillingStopCriterion.get()) < 0) {
+                optimalLeaf = leaf;
+                combinationFulfillingStopCriterion = Optional.of(networkActionCombination);
+            }
```
Comment (baptsegu) [inline ra-optimisation/search-tree-rao/src/main/java/com/farao_community/farao/search_tree_rao/SearchTree.java:409]: we only go in there `if (improvedEnough(leaf))`

Souldn't we put this outside of this if ?

Comment (baptsegu) [reply ra-optimisation/search-tree-rao/src/main/java/com/farao_community/farao/search_tree_rao/SearchTree.java:409]: ok, i just see you changed improvedEnough, nevermind. :p

**Review by baptsegu — CHANGES_REQUESTED** [review]

File: ra-optimisation/search-tree-rao/src/main/java/com/farao_community/farao/search_tree_rao/SearchTree.java (hunk 1 above)
Comment (phiedw) [reply ra-optimisation/search-tree-rao/src/main/java/com/farao_community/farao/search_tree_rao/SearchTree.java:409]: yeah I had to change the logic slightly to be able to check we respect min impacts, and still be able to update the optimalLeaf with a leaf with a higher cost that still reaches the stop criterion

File: ra-optimisation/search-tree-rao/src/main/java/com/farao_community/farao/search_tree_rao/SearchTree.java [hunk 2]
```diff
@@ -262,6 +271,10 @@ private void updateOptimalLeafWithNextDepthBestLeaf(AbstractNetworkPool networkP
         }
     }
 
+    private int orderNetworkActionCombinationsRandomly(NetworkActionCombination ra1, NetworkActionCombination ra2) {
```
[the remark this answers is not in the document]
Comment (phiedw) [reply ra-optimisation/search-tree-rao/src/main/java/com/farao_community/farao/search_tree_rao/SearchTree.java:274]: very well, I'll change it for the rangeActions too

File: ra-optimisation/search-tree-rao/src/main/java/com/farao_community/farao/search_tree_rao/SearchTree.java [hunk 3]
```diff
@@ -373,9 +390,23 @@ private SensitivityComputer getSensitivityComputerForOptimizationBasedOn(FlowRes
...
 
-    private synchronized void updateOptimalLeaf(Leaf leaf) {
+    private synchronized void updateOptimalLeaf(Leaf leaf, NetworkActionCombination networkActionCombination) {
         if (improvedEnough(leaf)) {
-            optimalLeaf = leaf;
+            // nominal case: stop criterion hasn't been reached yet
+            if (combinationFulfillingStopCriterion.isEmpty() && leaf.getCost() < optimalLeaf.getCost()) {
+                optimalLeaf = leaf;
+                if (stopCriterionReached(leaf)) {
+                    topLevelLogger.info("Stop criterion reached, other threads may skip optimization.");
```
[the remark this answers is not in the document]
Comment (phiedw) [reply ra-optimisation/search-tree-rao/src/main/java/com/farao_community/farao/search_tree_rao/SearchTree.java:399]: ok!

**Pull request merged**